### PR TITLE
[luxtronikheatpump] Fix control signal circulating pump

### LINKED ISSUE #15836: [luxtronikheatpump] Failed to update item ControlSignalCirculatingPump after upgrading to OH4
## Current Behavior
After upgrading from OH3 to OH4, the following messages are now appearing in the log file:
`15:12:28.103 [WARN ] [openhab.core.library.items.NumberItem] - Failed to update item 'HeatpumpwithLuxtronikControl_ControlSignalCirculatingPump' because '0 %' could not be converted to the item unit 'kWh'
`

I think it may be related to the definition here:
https://github.com/openhab/openhab-addons/blob/59c3135ccb911f9c9225c4f449e23b7f09e56ffb/bundles/org.openhab.binding.luxtronikheatpump/src/main/resources/OH-INF/thing/channels.xml#L1578

Any ideas if this may be the case or how to fix this otherwise?

### PR DESCRIPTION
**Pull request by jamietownsend**
Fixes #15836

### CODE REVIEW

File: bundles/org.openhab.binding.luxtronikheatpump/src/main/java/org/openhab/binding/luxtronikheatpump/internal/enums/HeatpumpChannel.java [hunk 1]
```diff
@@ -938,7 +938,7 @@ public enum HeatpumpChannel {
      * Control signal circulating pump
      * (original: Steuersignal Umwälzpumpe)
      */
-    CHANNEL_HZIO_PWM(183, "controlSignalCirculatingPump", NumberItem.class, Units.PERCENT, false, null),
+    CHANNEL_HZIO_PWM(183, "controlSignalCirculatingPump", NumberItem.class, null, false, null),
```
Comment (sgiehl) [inline bundles/org.openhab.binding.luxtronikheatpump/src/main/java/org/openhab/binding/luxtronikheatpump/internal/enums/HeatpumpChannel.java:941]: We need to keep the percent unit here. Based on the [LoxWiki](https://loxwiki.atlassian.net/wiki/spaces/LOX/pages/1533935933/Java+Webinterface) this channel provides a percent value, and the provided value needs to be divided by 10. This happens automatically for percent values, but won't happen without the unit. See https://github.com/openhab/openhab-addons/blob/dc3ca47b1218747f2b3770a3d7a899f5f91b8e05/bundles/org.openhab.binding.luxtronikheatpump/src/main/java/org/openhab/binding/luxtronikheatpump/internal/ChannelUpdaterJob.java#L135-L152

File: bundles/org.openhab.binding.luxtronikheatpump/src/main/resources/OH-INF/thing/channels.xml [hunk 2]
```diff
@@ -1572,10 +1572,9 @@
 	</channel-type>
 
 	<channel-type id="controlSignalCirculatingPump" advanced="true">
-		<item-type>Number:Energy</item-type>
+		<item-type>Number</item-type>
 		<label>Control Signal Circulating Pump</label>
-		<category>Energy</category>
-		<state pattern="%.1f %%" readOnly="true"/>
+		<state pattern="%.1f" readOnly="true"/>
```
Comment (sgiehl) [inline bundles/org.openhab.binding.luxtronikheatpump/src/main/resources/OH-INF/thing/channels.xml:1577]: We can keep the percent sign here.
```suggestion
		<state pattern="%.1f %%" readOnly="true"/>
```

File: bundles/org.openhab.binding.luxtronikheatpump/README.md [hunk 3]
```diff
@@ -207,7 +207,7 @@ The following channels are holding read only values:
 | highPressure | Number:Pressure | x | High pressure |
 | lowPressure | Number:Pressure | x | Low pressure |
 | outputCompressorHeating | Switch | x | Output compressor heating |
-| controlSignalCirculatingPump | Number:Energy | x | Control signal circulating pump |
+| controlSignalCirculatingPump | Number | x | Control signal circulating pump |
```
Comment (sgiehl) [inline bundles/org.openhab.binding.luxtronikheatpump/README.md:210]: `Number` seems fine to me, but guess `Number:Dimensionless` for percent values would be "more" correct. See https://www.openhab.org/docs/concepts/units-of-measurement.html

**Review by sgiehl — CHANGES_REQUESTED** [review]
Hey @jamietownsend 
Thanks for providing this pull request. The Energy definition was for sure added by accident. 

I'm actually not using this channel on my heatpump as it doesn't provide a value in my set up. Otherwise I might have noticed it.

I've added some comments, as some changes might be required to have that working correctly.

File: bundles/org.openhab.binding.luxtronikheatpump/README.md (hunk 3 above)
Comment (jamietownsend) [reply bundles/org.openhab.binding.luxtronikheatpump/README.md]: Agreed. Updated description from https://mw.ait-group.net/files/docs/EN/A0220/83055400.pdf

File: bundles/org.openhab.binding.luxtronikheatpump/src/main/java/org/openhab/binding/luxtronikheatpump/internal/enums/HeatpumpChannel.java (hunk 1 above)
Comment (jamietownsend) [reply bundles/org.openhab.binding.luxtronikheatpump/src/main/java/org/openhab/binding/luxtronikheatpump/internal/enums/HeatpumpChannel.java]: Changed back to Units.PERCENT

File: bundles/org.openhab.binding.luxtronikheatpump/src/main/resources/OH-INF/thing/channels.xml [hunk 4]
```diff
@@ -1572,9 +1572,8 @@
 	</channel-type>
 
 	<channel-type id="controlSignalCirculatingPump" advanced="true">
-		<item-type>Number:Energy</item-type>
+		<item-type>Number:Dimensionless</item-type>
 		<label>Control Signal Circulating Pump</label>
```
Comment (jlaur) [inline bundles/org.openhab.binding.luxtronikheatpump/src/main/resources/OH-INF/thing/channels.xml:1576]: Should this label be updated as well, like you did in the README? The README version is too long for a label, so a description could also help.

**Review by jlaur — CHANGES_REQUESTED** [review]
Thanks for the fix! Indeed this is the correct dimension which is used for updating the channel:
https://github.com/openhab/openhab-addons/blob/24f4234e9b316db801f94eaeb0c9710f8e850000/bundles/org.openhab.binding.luxtronikheatpump/src/main/java/org/openhab/binding/luxtronikheatpump/internal/enums/HeatpumpChannel.java#L937-L941

Since the channel type changed, you need to provide update instructions for the channel. See https://www.openhab.org/docs/developer/bindings/thing-xml.html#updating-thing-types

A similar example can be found here:
https://github.com/openhab/openhab-addons/blob/24f4234e9b316db801f94eaeb0c9710f8e850000/bundles/org.openhab.binding.miele/src/main/resources/OH-INF/update/instructions.xml#L11-L13

There is a slight doubt for me though. I see statically declared channels:
https://github.com/openhab/openhab-addons/blob/24f4234e9b316db801f94eaeb0c9710f8e850000/bundles/org.openhab.binding.luxtronikheatpump/src/main/resources/OH-INF/thing/thing-types.xml#L168

But also code that seems to dynamically add channels:
https://github.com/openhab/openhab-addons/blob/24f4234e9b316db801f94eaeb0c9710f8e850000/bundles/org.openhab.binding.luxtronikheatpump/src/main/java/org/openhab/binding/luxtronikheatpump/internal/LuxtronikHeatpumpHandler.java#L274-L293

The update instructions are not needed if the channel is actually added dynamically.

You can probably tell me where I am wrong. 🙂 But otherwise you could test this by first running the bundled version of the binding. Then create a managed Thing using the UI. Then uninstall the binding and drop your JAR with the fix. Then verify if the channel type is automatically updated after reinitialization, so that a new item will automatically be suggested as `Number:Dimensionless`. If it does, it looks like the channel is created dynamically. If not, most likely update instructions are needed. In this case you could add them and run the test again.

**Comment by jamietownsend** [comment]
> Thanks for the fix! Indeed this is the correct dimension which is used for updating the channel:
> 
> https://github.com/openhab/openhab-addons/blob/24f4234e9b316db801f94eaeb0c9710f8e850000/bundles/org.openhab.binding.luxtronikheatpump/src/main/java/org/openhab/binding/luxtronikheatpump/internal/enums/HeatpumpChannel.java#L937-L941
> 
> Since the channel type changed, you need to provide update instructions for the channel. See https://www.openhab.org/docs/developer/bindings/thing-xml.html#updating-thing-types
> 
> A similar example can be found here:
> 
> https://github.com/openhab/openhab-addons/blob/24f4234e9b316db801f94eaeb0c9710f8e850000/bundles/org.openhab.binding.miele/src/main/resources/OH-INF/update/instructions.xml#L11-L13
> 
> There is a slight doubt for me though. I see statically declared channels:
> 
> https://github.com/openhab/openhab-addons/blob/24f4234e9b316db801f94eaeb0c9710f8e850000/bundles/org.openhab.binding.luxtronikheatpump/src/main/resources/OH-INF/thing/thing-types.xml#L168
> 
> But also code that seems to dynamically add channels:
> 
> https://github.com/openhab/openhab-addons/blob/24f4234e9b316db801f94eaeb0c9710f8e850000/bundles/org.openhab.binding.luxtronikheatpump/src/main/java/org/openhab/binding/luxtronikheatpump/internal/LuxtronikHeatpumpHandler.java#L274-L293
> 
> The update instructions are not needed if the channel is actually added dynamically.
> 
> You can probably tell me where I am wrong. 🙂 But otherwise you could test this by first running the bundled version of the binding. Then create a managed Thing using the UI. Then uninstall the binding and drop your JAR with the fix. Then verify if the channel type is automatically updated after reinitialization, so that a new item will automatically be suggested as `Number:Dimensionless`. If it does, it looks like the channel is created dynamically. If not, most likely update instructions are needed. In this case you could add them and run the test again.

I'm not sure what needs to be done here. After following your suggested approach, I see that the Channel definition under the Thing shows "(Number:Dimensionless)", but the Item shows "Number:Energy". Deleting the Item and recreating it doesn't fix the problem. I notice the following while adding the Item from the Thing:
Screenshot 2023-11-08 at 12 20 08

Notice that the types are not the same. I also notice that the file .../openhab/userdata/jsondb/org.openhab.core.thing.Thing.json doesn't contain the correct value, even though the channel appears right on the Thing in the UI.

Deleting the Item and the Thing and then re-creating them solves the conflict.

@jlaur, any ideas how to proceed? (@sgiehl, FYI)

**Comment by jamietownsend** [comment]
I added an update.xml to delete and re-create the controlSignalCirculatingPump channel. This enables the correct creation of a new Item. Old items will still have issues and will have to be deleted and re-created.

**Comment by jlaur** [comment]
> I added an update.xml to delete and re-create the controlSignalCirculatingPump channel. This enables the correct creation of a new Item. Old items will still have issues and will have to be manually deleted and re-created.

Yes, the update instructions will only update the channel definitions in jsondb, but will touch linked items.

Did you notice a difference between having the update instructions when upgrading the binding and not having them? Since I'm confused about whether this binding provide static or dynamic channels (or both), I would like to figure out if the instructions are needed in this case.

**Comment by jlaur** [comment]
> Did you notice a difference between having the update instructions when upgrading the binding and not having them? Since I'm confused about whether this binding provide static or dynamic channels (or both), I would like to figure out if the instructions are needed in this case.

If the channel was automatically updated for your existing Thing when you updated the binding, the instructions are not needed. You have not added the `thingTypeVersion` property, so the instructions haven't been triggered.

**Comment by jamietownsend** [comment]
> If the channel was automatically updated for your existing Thing when you updated the binding, the instructions are not needed.

This is the case. The Thing was updated automatically, even without the update.xml.

> You have not added the `thingTypeVersion` property, so the instructions haven't been triggered.

Thanks for this clarification. I tested what happened when the `thingTypeVersion` property was correctly configured and I saw in the log that the update was correctly triggered and the thingTypeVersion changed from 0 to 1, but otherwise, there was no difference in what was changed. The Thing was updated, but the Item was not.

As such I'll remove the update.xml again to keep things simple.

Question: is there any way that we can automatically have the Items updated or should we write something in the docs to explain what has to be done?

**Comment by jamietownsend** [comment]
Not sure why this was closed. I'll try to re-open it...

**Comment by jamietownsend** [comment]
Ok, re-opening 🙂

**Comment by jlaur** [comment]
> Question: is there any way that we can automatically have the Items updated or should we write something in the docs to explain what has to be done?

There is no way to automatically update the items, but you could add a note somewhere around here:

https://github.com/openhab/openhab-distro/blob/b431198b21b0f435f62f5ebe671007dfafa7b53a/distributions/openhab/src/main/resources/bin/update.lst#L120-L122

These messages will be displayed during an upgrade.

**Comment by jamietownsend** [comment]
Thanks @jlaur. See related pull request: https://github.com/openhab/openhab-distro/pull/1606

I hope both pull requests can now me merged 🙂

**Review by sgiehl — APPROVED** [review]
To me this looks good now. 

Note: I'm unable to say if the new label matches what the channel actually provides. The previously used label was defined by "reverse engineering" and looking up what the java applet on the heatpump was using for it.

**Review by jlaur — APPROVED** [review]
LGTM

**Pull request merged**